### PR TITLE
allow packages to have internal dependencies

### LINKED ISSUE #240: initial import fails if python packages are used, but python is not installed
It looks like "python" is called by autobuild before the osdeps are installed

```
building initial autoproj import log, this may take a while
/usr/lib/ruby/2.3.0/open3.rb:199:in `spawn': No such file or directory - python (Errno::ENOENT)
	from /usr/lib/ruby/2.3.0/open3.rb:199:in `popen_run'
	from /usr/lib/ruby/2.3.0/open3.rb:95:in `popen3'
	from /root/.local/share/autoproj/gems/ruby/2.3.0/gems/autobuild-1.17.0/lib/autobuild/packages/python.rb:53:in `python_path'
	from /root/.local/share/autoproj/gems/ruby/2.3.0/gems/autobuild-1.17.0/lib/autobuild/packages/python.rb:89:in `update_environment'
```

### PR DESCRIPTION
**Pull request by g-arjones**
Fixes #240 

@doudou Thoughts?
@planthaber Could you please test if this fixes the issue you opened?

### CODE REVIEW

**Comment by planthaber** [comment]
Hi,

Yes, it fixes the issue.

Btw.: I saw that the import_package also has no dependencies yet. (none of wget,curl,tar,unzip) Bit i don't know what is really used.

https://github.com/rock-core/autoproj/blob/c8710396951e7a45c28b496e54fd7c525b459fc6/lib/autoproj/autobuild_extensions/dsl.rb#L163

**Comment by g-arjones** [comment]
import_package is a "generic" package, the dependencies must be defined by the user in the package definition:

```ruby
import_package 'package' do |pkg|
    pkg.depends_on 'dep1'
end
```

**Comment by planthaber** [comment]
But import_package uses the autobuild::Archive importer right?, which depends on tar,zip, etc?

But i think it won't be a similar issue, as tar should be installed by default is most distributions

**Comment by g-arjones** [comment]
Ah, I get what you mean now.. This are actually importer dependencies, not package-type dependencies. You could, for example, have a `cmake_package` that depends on wget and tar if you have something like:

In an *.autobuild file:
```
cmake_package 'package'
```

And in a source.yml:

```yml
  - package:
    type: archive
    url: ftp://example.com/package.tar.gz
```

Similarly, you could have an `importer_package` that do not depend on either if the VCS definition type is just `git`.. Anyway, the archive importer does have its dependencies installed automatically by autoproj based on the vcs type: https://github.com/rock-core/autoproj/blob/ebc72e17e146d8083ee96060a8da9f2cfd86082c/lib/autoproj/ops/import.rb#L129 

The actual packages that will get installed are defined here: https://github.com/rock-core/autoproj/blob/ebc72e17e146d8083ee96060a8da9f2cfd86082c/lib/autoproj/default.osdeps#L131

Maybe we should add `wget` (on linux) and `curl (bsd)` there...

**Comment by doudou** [comment]
> Maybe we should add `wget` (on linux) and `curl` (bsd) there...

No need, the archive importer now downloads using a pure ruby implementation.

File: lib/autoproj/autobuild_extensions/dsl.rb [hunk 1]
```diff
@@ -164,8 +164,11 @@ def import_package(name, workspace: Autoproj.workspace, &block)
     package_common(:import, name, workspace: Autoproj.workspace, &block)
 end
 
-def python_package(name, workspace: Autoproj.workspace, &block)
-    package_common(:python, name, workspace: Autoproj.workspace, &block)
+def python_package(name, workspace: Autoproj.workspace)
+    package_common(:python, name, workspace: Autoproj.workspace) do |pkg|
+        pkg.internal_dependencies << 'python'
```
Comment (doudou) [inline lib/autoproj/autobuild_extensions/dsl.rb]: I'm not sold on the `internal_dependencies` name. These are basically dependencies required during import, right ? So what about `import_dependencies` or `early_dependencies` ? Something more descriptive of how they're different from the "normal" dependencies.

File: lib/autoproj/autobuild_extensions/package.rb [hunk 2]
```diff
@@ -3,7 +3,7 @@ module AutobuildExtensions
         module Package
             # Tags explicitely added with #add_tag
             attr_reader :added_tags
-
+            attr_reader :internal_dependencies
```
Comment (doudou) [inline lib/autoproj/autobuild_extensions/package.rb]: I'm moving away from my very personal style of defining accessors for all instance variables. Even if you do add it, I'd prefer having an `add_` accessor for it.

**Comment by g-arjones** [comment]
> the archive importer now downloads using a pure ruby implementation

Not through ftp...

**Comment by doudou** [comment]
> Not through ftp...

True ...

File: lib/autoproj/autobuild_extensions/dsl.rb (hunk 1 above)
Comment (g-arjones) [reply lib/autoproj/autobuild_extensions/dsl.rb]: > These are basically dependencies required during import, right?

Not really.. In this particular case, it required for settings up the environment for python source pacakges (so, after import). The general use case I guess is for dependencies of the `Autobuild::Packages` subclasses (autobuild's own dependencies, in a sense. which is why I suggested internal)

Comment (g-arjones) [reply lib/autoproj/autobuild_extensions/dsl.rb]: Also, I don't think it is a good idea to advertise this too much as there are caveats (we assume that these dependencies are always osdeps and that they do not depend on strict package managers). So, generally, the regular user should not use this

File: lib/autoproj/autobuild_extensions/package.rb (hunk 2 above)
Comment (g-arjones) [reply lib/autoproj/autobuild_extensions/package.rb]: Sounds good to me...

File: lib/autoproj/autobuild_extensions/dsl.rb (hunk 1 above)
Comment (doudou) [reply lib/autoproj/autobuild_extensions/dsl.rb]: Fair enough ... Good for `internal` then. Mark them explicitly as `@api private` then.

**Comment by g-arjones** [comment]
Updated

**Comment by doudou** [comment]
@planthaber could you test master now ? I'll do a bugfix release afterwards.

**Pull request merged**